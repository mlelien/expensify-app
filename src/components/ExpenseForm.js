import React from "react";
import moment from "moment";
import PropTypes from "prop-types";
import { SingleDatePicker } from "react-dates";
import { expensePropTypes } from "../lib/proptypes";

export default class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      description: props.expense ? props.expense.description : "",
      note: props.expense ? props.expense.note : "",
      amount: props.expense ? (props.expense.amount / 100).toString() : "",
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      error: ""
    };
  }

  onDescriptionChange = e => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  };

  onNoteChange = e => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  };

  onAmountChange = e => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  };

  onDateChange = createdAt => {
    if (createdAt) {
      this.setState(() => ({ createdAt }));
    }
  };

  onCalendarFocusChanged = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }));
  };

  onSubmit = e => {
    const { amount, description, createdAt, note } = this.state;
    const { onSubmit } = this.props;
    e.preventDefault();

    if (!amount || !description) {
      this.setState(() => ({ error: "Please type in amount and description" }));
    } else {
      this.setState(() => ({ error: "" }));

      onSubmit({
        description,
        amount: parseFloat(amount, 10) * 100,
        createdAt: createdAt.valueOf(),
        note
      });
    }
  };

  render() {
    const {
      error,
      amount,
      description,
      calendarFocused,
      value,
      createdAt
    } = this.state;

    return (
      <div>
        {error && <p className='form__error'>{error}</p>}
        <form className='form' onSubmit={this.onSubmit}>
          <input
            className='text-input'
            type='text'
            placeholder='Description'
            value={description}
            onChange={this.onDescriptionChange}
          />
          <input
            className='text-input'
            type='text'
            placeholder='Amount'
            value={amount}
            onChange={this.onAmountChange}
          />
          <textarea
            className='textarea'
            value={value}
            onChange={this.onNoteChange}
            placeholder='Note (optional)'
          />
          <SingleDatePicker
            date={createdAt}
            onDateChange={this.onDateChange}
            focused={calendarFocused}
            onFocusChange={this.onCalendarFocusChanged}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <div>
            <button className='button' type='submit'>Save Expense</button>
          </div>
        </form>
      </div>
    );
  }
}

ExpenseForm.defaultProps = {
  expense: undefined
};

ExpenseForm.propTypes = {
  onSubmit: PropTypes.func.isRequired,
  expense: expensePropTypes
};
