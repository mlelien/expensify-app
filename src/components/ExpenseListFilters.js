import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import PropTypes from 'prop-types';
import {
  setTextFilter,
  sortByDate,
  sortByAmount,
  setStartDate,
  setEndDate
} from '../actions/filters';

export class ExpenseListFilters extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      focusedInput: null
    };
  }

  onDatesChange = ({ startDate, endDate }) => {
    const { setStartDate: setStartDateLocal, setEndDate: setEndDateLocal } = this.props;

    setStartDateLocal(startDate);
    setEndDateLocal(endDate);
  };

  onFocusChange = focusedInput => {
    this.setState(() => ({ focusedInput }));
  };

  textFilterOnChange = e => {
    const { setTextFilter: setTextFilterLocal } = this.props;
    setTextFilterLocal(e.target.value);
  };

  selectFilterOnChange = e => {
    const { sortByDate: sortByDateLocal, sortByAmount: sortByAmountLocal } = this.props;
    if (e.target.value === 'date') {
      sortByDateLocal();
    } else if (e.target.value === 'amount') {
      sortByAmountLocal();
    }
  };

  render() {
    const {
      filters: { text, sortBy, startDate, endDate }
    } = this.props;

    const { focusedInput } = this.state;

    return (
      <div className='content-container'>
        <div className='input-group'>
          <div className='input-group__item'>
            <input 
              className='text-input' 
              type='text' 
              value={text} 
              placeholder='Search expenses'
              onChange={this.textFilterOnChange} />
          </div>
          <div className='input-group__item'>
            <select className='select' value={sortBy} onChange={this.selectFilterOnChange}>
              <option value='date'>Date</option>
              <option value='amount'>Amount</option>
            </select>
          </div>
          <div className='input-group__item'>
            <DateRangePicker
              startDate={startDate}
              startDateId='startDate'
              endDate={endDate}
              endDateId='endDate'
              onDatesChange={this.onDatesChange}
              focusedInput={focusedInput}
              onFocusChange={this.onFocusChange}
              numberOfMonths={1}
              isOutsideRange={() => false}
              showClearDates
            />
          </div>
        </div>
      </div>
    );
  }
}

ExpenseListFilters.defaultProps = {
  filters: undefined
};

ExpenseListFilters.propTypes = {
  setStartDate: PropTypes.func.isRequired,
  setEndDate: PropTypes.func.isRequired,
  setTextFilter: PropTypes.func.isRequired,
  sortByAmount: PropTypes.func.isRequired,
  sortByDate: PropTypes.func.isRequired,
  filters: PropTypes.shape({
    text: PropTypes.string.isRequired,
    sortBy: PropTypes.string.isRequired,
    startDate: PropTypes.object,
    endDate: PropTypes.object
  })
};

const mapStateToProps = state => ({
  filters: state.filters
});

const mapDispatchToProps = dispatch => ({
  setStartDate: startDate => dispatch(setStartDate(startDate)),
  setEndDate: endDate => dispatch(setEndDate(endDate)),
  setTextFilter: value => dispatch(setTextFilter(value)),
  sortByDate: () => dispatch(sortByDate()),
  sortByAmount: () => dispatch(sortByAmount())
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ExpenseListFilters);
