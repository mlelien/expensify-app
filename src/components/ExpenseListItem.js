import React from "react";
import { NavLink } from "react-router-dom";
import PropTypes from "prop-types";
import moment from 'moment';
import numeral from 'numeral';

const ExpenseListItem = ({ id, description, amount, createdAt }) => (
  <div>
    <NavLink exact to={`/edit/${id}`}><h3>{description}</h3></NavLink>
    <p>
      {numeral(amount / 100).format('$0,0.00')} 
      - 
      {moment(createdAt).format('MMMM Do, YYYY')}
    </p>
  </div>
);

ExpenseListItem.defaultProps = {
  createdAt: undefined
};

ExpenseListItem.propTypes = {
  id: PropTypes.string.isRequired,
  description: PropTypes.string.isRequired,
  amount: PropTypes.number.isRequired,
  createdAt: PropTypes.number
};

export default ExpenseListItem;
