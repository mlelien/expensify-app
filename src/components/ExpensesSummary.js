import React from 'react';
import { connect } from "react-redux";
import numeral from 'numeral';
import { Link } from 'react-router-dom';
import PropTypes from "prop-types";
import getVisibleExpenses from "../selectors/getVisibleExpenses";
import getExpensesTotal from "../selectors/getExpensesTotal";

export const ExpensesSummary = (props) => {
  const { expensesCount, totalAmount } = props;
  const expenseWord = expensesCount === 1 ? 'expense' : 'expenses';
  const total = numeral(totalAmount / 100).format('$0,0.00');

  return (
    <div className='page-header'>
      <div className='content-container'>
        <h1 className='page-header__title'>
          Viewing <span>{expensesCount}</span> {expenseWord} totalling <span>{total}</span>
        </h1>
        <div className='page-header__actions'>
          <Link className='button' to='/create'>Add Expense</Link>
        </div>
      </div>
    </div>
  );
}

ExpensesSummary.propTypes = {
  expensesCount: PropTypes.number.isRequired,
  totalAmount: PropTypes.number.isRequired
};

const mapStateToProps = state => ({
  expensesCount: getVisibleExpenses(state.expenses, state.filters).length,
  totalAmount: getExpensesTotal(state.expenses, state.filters)
});

export default connect(mapStateToProps)(ExpensesSummary);