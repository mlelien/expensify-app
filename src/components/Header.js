import React from "react";
import { connect } from 'react-redux';
import { Link } from "react-router-dom";
import PropTypes from 'prop-types';
import { startLogout } from '../actions/auth';

export const Header = (props) => {
  const { startLogout: startLogoutLocal } = props;

  return (
    <header className='header'>
      <div className='content-container'>
        <div className='header__content'>
          <Link className='header__title' to='/dashboard'><h1>Expensify</h1></Link>
          <button className='button button--link' type='submit' onClick={startLogoutLocal}>Logout</button>
        </div>
      </div>
    </header>
  )
};

Header.propTypes = {
  startLogout: PropTypes.func.isRequired
};

const mapDispatchToProps = (dispatch) => ({
  startLogout: dispatch(startLogout)
});

export default connect(undefined, mapDispatchToProps)(Header);
