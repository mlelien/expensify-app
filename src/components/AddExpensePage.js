import React from "react";
import { connect } from "react-redux";
import PropTypes from "prop-types";
import { startAddExpense } from "../actions/expenses";
import ExpenseForm from "./ExpenseForm";
import { history } from '../routers/AppRouter';

export class AddExpensePage extends React.Component {
  onSubmit = expense => {
    const { startAddExpense: startAddExpenseLocal } = this.props;

    startAddExpenseLocal(expense);
    history.push("/dashboard");
  };

  render() {
    return (
      <div>
        <div className='page-header'>
          <div className='content-container'>
            <h1 className='page-header__title'>Add Expense</h1>
          </div>
        </div>
        <div className='content-container'>
          <ExpenseForm onSubmit={this.onSubmit} />
        </div>
      </div>
    );
  }
}

AddExpensePage.propTypes = {
  startAddExpense: PropTypes.func.isRequired,
  // history: PropTypes.shape({
  //   push: PropTypes.func.isRequired
  // }).isRequired
};

const mapDispatchToProps = dispatch => ({
  startAddExpense: expense => dispatch(startAddExpense(expense))
});

export default connect(
  undefined,
  mapDispatchToProps
)(AddExpensePage);
