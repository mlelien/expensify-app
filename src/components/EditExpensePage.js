import React from "react";
import { connect } from "react-redux";
import PropTypes from "prop-types";
import ExpenseForm from "./ExpenseForm";
import { startEditExpense, startRemoveExpense } from "../actions/expenses";
import { expensePropTypes } from "../lib/proptypes";
import { history } from '../routers/AppRouter';

export class EditExpensePage extends React.Component {
  onSubmit = expense => {
    const {
      startEditExpense: startEditExpenseLocal,
      expense: { id }
    } = this.props;
    startEditExpenseLocal(id, expense);
    history.push("/dashboard");
  };

  onRemove = () => {
    const {
      startRemoveExpense: startRemoveExpenseLocal,
      expense: { id },
    } = this.props;
    startRemoveExpenseLocal(id);
    history.push("/dashboard");
  };

  render() {
    const { expense } = this.props;
    return (
      <div>
        <ExpenseForm expense={expense} onSubmit={this.onSubmit} />
        <button type='button' onClick={this.onRemove}>
          Remove
        </button>
      </div>
    );
  }
}

EditExpensePage.propTypes = {
  startEditExpense: PropTypes.func.isRequired,
  startRemoveExpense: PropTypes.func.isRequired,
  expense: expensePropTypes.isRequired
};

const mapStateToProps = (state, props) => ({
  expense: state.expenses.find(expense => expense.id === props.computedMatch.params.id)
});

const mapDispatchToProps = dispatch => ({
  startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
  startRemoveExpense: id => dispatch(startRemoveExpense({ id }))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(EditExpensePage);
