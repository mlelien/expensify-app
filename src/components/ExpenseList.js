import React from "react";
import { connect } from "react-redux";
import PropType from "prop-types";
import { expensePropTypes } from "../lib/proptypes";
import getVisibleExpenses from "../selectors/getVisibleExpenses";
import ExpenseListItem from "./ExpenseListItem";

export const ExpenseList = props => {
  const { expenses } = props;

  return (
    <div className='content-container'>

      <div className='list-header'>
        <div className='show-for-mobile'>Expenses</div>
        <div className='show-for-desktop'>Expense</div>
        <div className='show-for-desktop'>Amount</div>
      </div>

      <div className='list-body'>
        {expenses.length === 0 ? (
          <div className='list-item__message'>
            <span>No expenses</span>
          </div>
        ) : (
          expenses.map(expense => (
            <ExpenseListItem key={expense.id} {...expense} />
          ))
        )}
      </div>
    </div>
  );
};

ExpenseList.propTypes = {
  expenses: PropType.arrayOf(expensePropTypes).isRequired
};

const mapStateToProps = state => ({
  expenses: getVisibleExpenses(state.expenses, state.filters)
});

export default connect(mapStateToProps)(ExpenseList);
