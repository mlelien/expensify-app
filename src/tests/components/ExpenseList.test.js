import React from "react";
import { shallow } from "enzyme";
import expenses from "../fixtures/expenses";
import { ExpenseList } from "../../components/ExpenseList";

test("ExpenseList with expenses", () => {
  const wrapper = shallow(<ExpenseList expenses={expenses} />);
  expect(wrapper).toMatchSnapshot();
});

test("ExpenseList - empty", () => {
  const wrapper = shallow(<ExpenseList expenses={[]} />);
  expect(wrapper).toMatchSnapshot();
});
