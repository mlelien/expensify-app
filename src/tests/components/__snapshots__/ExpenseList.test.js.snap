// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExpenseList - empty 1`] = `
<div
  className="content-container"
>
  <div
    className="list-header"
  >
    <div
      className="show-for-mobile"
    >
      Expenses
    </div>
    <div
      className="show-for-desktop"
    >
      Expense
    </div>
    <div
      className="show-for-desktop"
    >
      Amount
    </div>
  </div>
  <div
    className="list-body"
  >
    <div
      className="list-item__message"
    >
      <span>
        No expenses
      </span>
    </div>
  </div>
</div>
`;

exports[`ExpenseList with expenses 1`] = `
<div
  className="content-container"
>
  <div
    className="list-header"
  >
    <div
      className="show-for-mobile"
    >
      Expenses
    </div>
    <div
      className="show-for-desktop"
    >
      Expense
    </div>
    <div
      className="show-for-desktop"
    >
      Amount
    </div>
  </div>
  <div
    className="list-body"
  >
    <ExpenseListItem
      amount={4000}
      createdAt={0}
      description="Gas bill"
      id="1"
      key="1"
      note="Utilities"
    />
    <ExpenseListItem
      amount={125}
      createdAt={-345600000}
      description="Gum"
      id="2"
      key="2"
      note=""
    />
    <ExpenseListItem
      amount={225}
      createdAt={345600000}
      description="Coffee"
      id="3"
      key="3"
      note=""
    />
  </div>
</div>
`;
